doc/proposals/scorecard-user-tests.md

```diff
@@ -0,0 +1,168 @@
+start with `scorecard_function_`, we do a simple match (like `status/readyReplicas: 4`).
+
+This design would allow us to replace the old "Operator actions are reflected in status" (which would be tested by the `expected/status` check) and
+"Writing into CRs has an effect" (which would be tested by the `expected/resources` check) tests.
+
+### Plugin System
+
+In order to increase the flexibility of the user defined tests and allow users to implement more complex E2E style tests for scorecard,
+the user-defined tests will be implemented via a plugin system. Users will specify the path of the script they wish to run as well
+as environment variable to set for the command. The command would then print out the result as JSON to stdout. Here is an example:
+
+```yaml
+user_defined_tests:
+- path: "scorecard/simple-scorecard.sh"
+  env:
+    - CONFIG_FILE: "scorecard/simple-scorecard.yaml"
+- path: "scorecard/go-test.sh"
+  env:
+    - TEST_DIR: "./test/e2e"
+    - ENABLE_SCORECARD: true
+    - NAMESPACED_MANIFEST: "deploy/namespaced_init.yaml"
+    - GO_TEST_FLAGS: "-parallel=1"
+```
+
+The above is an example of a user-defined test suite with 2 tests: the new simple scorecard tests described above and a test
+built using the Operator SDK's test framework that's been modified to be able to output results in the standardized JSON output.
+
+Below is an example of what the JSON output of a test would look like:
+
+```json
+{
+    "error": 0,
+    "pass": 1,
+    "partial_pass": 1,
+    "fail": 0,
+    "total_tests": 2,
+    "total_score_percent": 71,
+    "tests": [
+        {
+            "state": "partial_pass",
+            "name": "Operator Actions Reflected In Status",
+            "description": "The operator updates the Custom Resources status when the application state is updated",
+            "earned": 2,
+            "maximum": 3,
+            "suggestions": [
+                {
+                    "suggestion": "Operator should update status when scaling cluster down"
+                }
+            ],
+            "errors": []
+        },
+        {
+            "state": "pass",
+            "name": "Verify health of cluster",
+            "description": "The cluster created by the operator is working properly",
+            "earned": 1,
+            "maximum": 1,
+            "suggestions": [],
+            "errors": []
+        }
+    ]
+}
+```
+
+This JSON output would make it simple for others to create scorecard plugins while keeping it simple for the scorecard
+to parse and integrate with the other tests. The above JSON design is based on the `TestResult` type with the addition
+of a `state` type, which can be `pass` (earned == max), `partial_pass` (0 < earned < max), `fail` (earned == 0), or `error`
+(fatal error; disregard output score). We also print the number of tests in each state as well as the total score to make
+it easier for users to quickly see what states the tests are in.
```

Comment:
Are there any good examples of other projects with a similar plugin system, specifically that expect the plugin to return data in a specific format? I ask because I wonder if we should version the interface so that we can make changes to it in the future without breaking old plugins.

Maybe a totally off the wall idea, but would it make sense to re-use the kubernetes API for this? Similar to how the [admission webhook request/response API](https://github.com/kubernetes/api/blob/3c12c96769cc4f6df2245f283ee74992c94fa4ee/admission/v1beta1/types.go#L29-L37) works? (btw, not suggesting we change to an HTTP interface or anything, just marshal things as k8s-compliant API objects)

Maybe something like:

```go
type ScorecardTest struct {
	metav1.TypeMeta `json:",inline"`
	// Spec describes the attributes for the test
	Spec *ScorecardTestSpec `json:"spec"`

	// Results describes the results of running the test.
	// +optional
	Results *ScorecardTestResults `json:"results,omitempty"`
}
```

And maybe even crazier, perhaps we could write an operator that could execute these tests in-cluster. Food for thought :slightly_smiling_face:

Reply:
Hmm. I like the idea because it allows for versioning, and this will most likely go through a couple of changes over time as we use this more and figure out what else we may need from the output. And reusing some of the other features of the API could be quite useful here.

Running the scorecard tests in-cluster shouldn't be too difficult either, assuming we have a lot of permissions and can actually create all the resources we need. So maybe that could be a runtime mode in the future (kind of like how we have `test local` and `test cluster`).